// axis-oracles/src/components/modules/CryptoModule.tsx
import { FC, useState, useMemo, useEffect } from 'react';
import { CalendarIcon, Loader2, CheckCircle, XCircle, Check, ChevronsUpDown, Clock } from 'lucide-react';
import bitcoinIcon from '@/assets/bitcoin-icon.png';
import { ModuleCard } from './ModuleCard';
import { Button } from '@/components/ui/button';
import { Label } from '@/components/ui/label';
import { Input } from '@/components/ui/input';
import { Calendar } from '@/components/ui/calendar';
import { Popover, PopoverContent, PopoverTrigger } from '@/components/ui/popover';
import { Command, CommandEmpty, CommandGroup, CommandInput, CommandItem, CommandList } from '@/components/ui/command';
import { FeeDisplay } from './FeeDisplay';
import { useWallet } from '@solana/wallet-adapter-react';
import { cn } from '@/lib/utils';
import { format } from 'date-fns';
import { fetchCryptoPrice, formatPrice, CryptoPriceData } from '@/utils/apiService';
import { toast } from '@/hooks/use-toast';
import { CRYPTO_LIST, CryptoAsset } from '@/data/cryptoList';
import { 
  getDefaultLocalResolutionDate, 
  getDefaultLocalHour, 
  getDefaultLocalMinute, 
  createDateTimeFromLocal, 
  formatDateTimeForTitle,
  formatLocalDateTime,
  isValidFutureTime,
  getTimezoneAbbr
} from '@/utils/utcTime';

import { DeploymentState } from '@/hooks/useCreateFeed';

interface CryptoModuleProps {
  onCreateFeed: (config: { 
    symbol: string; 
    quoteCurrency: 'USD';
    metric: 'price';
    resolutionDate: Date;
    title: string;
    logo?: string;
  }) => void;
  isLoading?: boolean;
  deploymentState?: DeploymentState;
  getButtonText?: (defaultText?: string) => string;
}

export const CryptoModule: FC<CryptoModuleProps> = ({ onCreateFeed, isLoading, deploymentState, getButtonText }) => {
  const { connected } = useWallet();
  const [selectedAsset, setSelectedAsset] = useState<CryptoAsset | null>(null);
  const [open, setOpen] = useState(false);
  
  // Initialize with smart local defaults (displayed as local, stored as UTC)
  const defaultDate = getDefaultLocalResolutionDate();
  const [resolutionDate, setResolutionDate] = useState<Date>(defaultDate);
  const [selectedHour, setSelectedHour] = useState<string>(getDefaultLocalHour(defaultDate));
  const [selectedMinute, setSelectedMinute] = useState<string>(getDefaultLocalMinute(defaultDate));
  const tzAbbr = getTimezoneAbbr();
  
  // Live preview state
  const [priceData, setPriceData] = useState<CryptoPriceData | null>(null);
  const [isFetching, setIsFetching] = useState(false);
  const [fetchError, setFetchError] = useState(false);

  // Calculate full resolution datetime (local input -> Date object with correct UTC)
  const fullResolutionDate = useMemo(() => {
    if (!resolutionDate) return undefined;
    return createDateTimeFromLocal(resolutionDate, selectedHour, selectedMinute);
  }, [resolutionDate, selectedHour, selectedMinute]);

  // Validate minimum 1 hour in future (UTC)
  const isValidTime = fullResolutionDate ? isValidFutureTime(fullResolutionDate) : true;

  // Fetch price data when asset changes
  useEffect(() => {
    const fetchData = async () => {
      if (!selectedAsset) {
        setPriceData(null);
        setFetchError(false);
        return;
      }

      setIsFetching(true);
      setFetchError(false);

      try {
        const data = await fetchCryptoPrice(selectedAsset.symbol);
        if (data) {
          setPriceData(data);
          setFetchError(false);
        } else {
          setPriceData(null);
          setFetchError(true);
        }
      } catch (error) {
        setPriceData(null);
        setFetchError(true);
        toast({
          title: "Failed to fetch data",
          description: "Could not retrieve price from GeckoTerminal",
          variant: "destructive",
        });
      } finally {
        setIsFetching(false);
      }
    };

    fetchData();
  }, [selectedAsset]);

  // Auto-generate feed title (shows UTC in title for global consistency)
  const autoGeneratedTitle = useMemo(() => {
    if (!selectedAsset || !fullResolutionDate) return '';
    const dateLabel = formatDateTimeForTitle(fullResolutionDate);
    return `${selectedAsset.symbol}/USD Price [${dateLabel}]`;
  }, [selectedAsset, fullResolutionDate]);

  const handleCreate = () => {
    if (!selectedAsset || !fullResolutionDate) return;
    onCreateFeed({
      symbol: selectedAsset.symbol,
      quoteCurrency: 'USD',
      metric: 'price',
      resolutionDate: fullResolutionDate,
      title: autoGeneratedTitle,
      logo: selectedAsset.logo,
    });
  };

  const canCreate = connected && selectedAsset && fullResolutionDate && priceData && isValidTime;

  return (
    <ModuleCard
      iconImage={bitcoinIcon}
      title="Global Crypto"
      description="Streaming high-precision data for global markets"
      index={0}
    >
      <div className="space-y-4 flex-1 flex flex-col">
        {/* Asset Selection Combobox */}
        <div className="space-y-2">
          <Label className="text-sm font-medium text-foreground">Asset</Label>
          <Popover open={open} onOpenChange={setOpen}>
            <PopoverTrigger asChild>
              <Button
                variant="outline"
                role="combobox"
                aria-expanded={open}
                className="w-full justify-between bg-background border-border max-w-full overflow-hidden"
              >
                {selectedAsset ? (
                  <div className="flex items-center gap-2 min-w-0 overflow-hidden">
                    <img 
                      src={selectedAsset.logo} 
                      alt={selectedAsset.symbol}
                      className="w-5 h-5 rounded-full shrink-0"
                      onError={(e) => { e.currentTarget.style.display = 'none'; }}
                    />
                    <span className="font-medium">{selectedAsset.symbol}</span>
                    <span className="text-muted-foreground truncate">({selectedAsset.name})</span>
                  </div>
                ) : (
                  <span className="text-muted-foreground truncate">Search crypto...</span>
                )}
                <ChevronsUpDown className="ml-2 h-4 w-4 shrink-0 opacity-50" />
              </Button>
            </PopoverTrigger>
            <PopoverContent className="w-[300px] p-0 bg-popover border-border z-50" align="start">
              <Command>
                <CommandInput placeholder="Search crypto..." />
                <CommandList>
                  <CommandEmpty>No cryptocurrency found.</CommandEmpty>
                  <CommandGroup className="max-h-[200px] overflow-y-auto">
                    {CRYPTO_LIST.map((crypto) => (
                      <CommandItem
                        key={crypto.symbol}
                        value={`${crypto.symbol} ${crypto.name}`}
                        onSelect={() => {
                          setSelectedAsset(crypto);
                          setOpen(false);
                        }}
                        className="flex items-center gap-2 cursor-pointer"
                      >
                        <img 
                          src={crypto.logo} 
                          alt={crypto.symbol}
                          className="w-5 h-5 rounded-full"
                          onError={(e) => { e.currentTarget.style.display = 'none'; }}
                        />
                        <span className="font-medium">{crypto.symbol}</span>
                        <span className="text-muted-foreground">({crypto.name})</span>
                        <Check
                          className={cn(
                            "ml-auto h-4 w-4",
                            selectedAsset?.symbol === crypto.symbol ? "opacity-100" : "opacity-0"
                          )}
                        />
                      </CommandItem>
                    ))}
                  </CommandGroup>
                </CommandList>
              </Command>
            </PopoverContent>
          </Popover>
          
          {/* Live Preview */}
          {isFetching && (
            <div className="flex items-center gap-2 text-sm text-muted-foreground">
              <Loader2 className="h-3 w-3 animate-spin" />
              <span>Fetching price...</span>
            </div>
          )}
          
          {priceData && !isFetching && selectedAsset && (
            <div className="flex items-center gap-2 p-2 bg-green-500/10 border border-green-500/20 rounded-md">
              <img 
                src={selectedAsset.logo} 
                alt={selectedAsset.symbol}
                className="w-5 h-5 rounded-full"
                onError={(e) => { e.currentTarget.style.display = 'none'; }}
              />
              <CheckCircle className="h-4 w-4 text-green-500 shrink-0" />
              <span className="text-sm">
                <span className="text-green-600 font-medium">{priceData.symbol}</span>
                <span className="text-muted-foreground"> | </span>
                <span className="text-foreground">Price: ${formatPrice(priceData.price)}</span>
              </span>
            </div>
          )}
          
          {fetchError && !isFetching && selectedAsset && (
            <div className="flex items-center gap-2 p-2 bg-destructive/10 border border-destructive/20 rounded-md">
              <XCircle className="h-4 w-4 text-destructive shrink-0" />
              <span className="text-sm text-destructive">Symbol not found on GeckoTerminal</span>
            </div>
          )}
        </div>

        {/* Metric Selector - Segmented Control */}
        <div className="space-y-2">
          <Label className="text-sm font-medium text-foreground">Metric to Track</Label>
          <div className="flex bg-muted rounded-lg p-1">
            <button
              type="button"
              className="flex-1 py-2 px-3 text-sm font-medium rounded-md transition-all duration-200 bg-primary text-primary-foreground shadow-sm"
            >
              Price ($)
            </button>
            <div
              className="flex-1 py-2 px-3 text-sm font-medium rounded-md text-muted-foreground/50 cursor-not-allowed flex items-center justify-center gap-1.5"
            >
              Volume
              <span className="text-[10px] uppercase font-bold bg-muted-foreground/20 px-1.5 py-0.5 rounded">Soon</span>
            </div>
          </div>
        </div>

        {/* Resolution Date & Time (Local) */}
        <div className="space-y-2">
          <Label className="text-sm font-medium text-foreground">
            Resolution Date & Time <span className="text-muted-foreground text-xs">({tzAbbr})</span>
          </Label>
          <div className="flex flex-wrap gap-2">
            <Popover>
              <PopoverTrigger asChild>
                <Button
                  variant="outline"
                  className={cn(
                    "flex-1 min-w-[140px] justify-start text-left font-normal bg-background border-border",
                    !resolutionDate && "text-muted-foreground"
                  )}
                >
                  <CalendarIcon className="mr-2 h-4 w-4 shrink-0" />
                  <span className="truncate">{resolutionDate ? format(resolutionDate, "PPP") : "Select date"}</span>
                </Button>
              </PopoverTrigger>
              <PopoverContent className="w-auto p-0 bg-popover border-border z-50" align="start" collisionPadding={16}>
                <Calendar
                  mode="single"
                  selected={resolutionDate}
                  onSelect={setResolutionDate}
                  disabled={(date) => date < new Date()}
                  initialFocus
                  className="p-3 pointer-events-auto"
                />
              </PopoverContent>
            </Popover>
            
            {/* Time Selection */}
            <div className="flex items-center gap-1">
              <Clock className="h-4 w-4 text-muted-foreground" />
              <Input
                type="text"
                inputMode="numeric"
                maxLength={2}
                value={selectedHour}
                onChange={(e) => {
                  const val = e.target.value.replace(/\D/g, '').slice(0, 2);
                  setSelectedHour(val);
                }}
                onBlur={(e) => {
                  let val = parseInt(e.target.value) || 0;
                  val = Math.min(23, Math.max(0, val));
                  setSelectedHour(val.toString().padStart(2, '0'));
                }}
                className="w-14 bg-background border-border text-center"
                placeholder="HH"
              />
              <span className="text-muted-foreground font-medium">:</span>
              <Input
                type="text"
                inputMode="numeric"
                maxLength={2}
                value={selectedMinute}
                onChange={(e) => {
                  const val = e.target.value.replace(/\D/g, '').slice(0, 2);
                  setSelectedMinute(val);
                }}
                onBlur={(e) => {
                  let val = parseInt(e.target.value) || 0;
                  val = Math.min(59, Math.max(0, val));
                  setSelectedMinute(val.toString().padStart(2, '0'));
                }}
                className="w-14 bg-background border-border text-center"
                placeholder="MM"
              />
            </div>
          </div>
          
          {/* Time Validation Error */}
          {fullResolutionDate && !isValidTime && (
            <p className="text-xs text-destructive flex items-center gap-1">
              <XCircle className="h-3 w-3" />
              Minimum duration is 2 minutes from now
            </p>
          )}
          
          <p className="text-xs text-muted-foreground">Select your local time — it will be converted to UTC automatically.</p>
        </div>

        {/* Auto-generated Feed Title Preview */}
        {autoGeneratedTitle && (
          <div className="space-y-2">
            <Label className="text-sm font-medium text-muted-foreground">Feed Title (auto)</Label>
            <div className="px-3 py-2 bg-muted/50 rounded-md border border-border/50 h-10 flex items-center">
              <span className="text-sm text-foreground truncate">{autoGeneratedTitle}</span>
            </div>
          </div>
        )}
        
        {/* Spacer to push button to bottom */}
        <div className="flex-1" />

        <div className="flex items-center justify-between pt-2">
          <FeeDisplay />
          <Button
            variant={deploymentState === 'success' ? 'default' : 'gold'}
            size="sm"
            onClick={handleCreate}
            disabled={!canCreate || isLoading}
            className={deploymentState === 'success' ? 'bg-green-500 hover:bg-green-600' : ''}
          >
            {getButtonText ? getButtonText('Create Oracle') : (isLoading ? 'Creating...' : 'Create Oracle')}
          </Button>
        </div>
      </div>
    </ModuleCard>
  );
};
