// axis-oracles/src/components/modules/EsportsModuleModal.tsx
import { FC, useState, useMemo, useEffect } from 'react';
import { CalendarIcon, Check, ChevronsUpDown, Search, Clock, Loader2, AlertCircle } from 'lucide-react';
import esportsIcon from '@/assets/esports-icon.png';
import { Button } from '@/components/ui/button';
import { Label } from '@/components/ui/label';
import { Popover, PopoverContent, PopoverTrigger } from '@/components/ui/popover';
import { Command, CommandEmpty, CommandGroup, CommandInput, CommandItem, CommandList } from '@/components/ui/command';
import { Dialog, DialogContent, DialogHeader, DialogTitle } from '@/components/ui/dialog';
import { FeeDisplay } from './FeeDisplay';
import { TransactionConfirmModal } from './TransactionConfirmModal';
import { useWallet } from '@solana/wallet-adapter-react';
import { cn } from '@/lib/utils';
import { format } from 'date-fns';
import { supabase } from '@/integrations/supabase/client';
import { DeploymentState } from '@/hooks/useCreateFeed';
import { addUTCHours, formatUTCDateShort, formatUTCTime } from '@/utils/utcTime';

type GameType = 'cs2' | 'dota2' | 'lol';

interface EsportsMatch {
  id: string;
  game: GameType;
  team1: string;
  team1Short: string;
  team2: string;
  team2Short: string;
  tournament: string;
  scheduledAt: string;
  status: string;
}

const getGameLabel = (game: GameType): string => {
  switch (game) {
    case 'cs2': return 'CS2';
    case 'dota2': return 'Dota 2';
    case 'lol': return 'LoL';
    default: return game;
  }
};

const getGameBadgeClasses = (game: GameType): string => {
  switch (game) {
    case 'cs2':
      return 'bg-orange-500/20 text-orange-400 border-orange-500/30';
    case 'dota2':
      return 'bg-red-500/20 text-red-400 border-red-500/30';
    case 'lol':
      return 'bg-blue-500/20 text-blue-400 border-blue-500/30';
    default:
      return 'bg-muted text-muted-foreground border-border';
  }
};

interface EsportsModuleModalProps {
  open: boolean;
  onOpenChange: (open: boolean) => void;
  onCreateFeed: (config: { 
    game: GameType;
    gameName: string;
    matchId: string;
    marketType: 'match_winner';
    resolutionDate: Date;
    title: string;
    team1Name?: string;
    team2Name?: string;
  }) => void;
  isLoading?: boolean;
  deploymentState?: DeploymentState;
  getButtonText?: (defaultText?: string) => string;
}

export const EsportsModuleModal: FC<EsportsModuleModalProps> = ({ 
  open, 
  onOpenChange, 
  onCreateFeed, 
  isLoading, 
  deploymentState, 
  getButtonText 
}) => {
  const { connected } = useWallet();
  const [selectedMatch, setSelectedMatch] = useState<EsportsMatch | null>(null);
  const [matchSearchOpen, setMatchSearchOpen] = useState(false);
  const [searchQuery, setSearchQuery] = useState('');
  const [resolutionDate, setResolutionDate] = useState<Date>();
  const [showConfirmModal, setShowConfirmModal] = useState(false);
  
  const [matches, setMatches] = useState<EsportsMatch[]>([]);
  const [isFetchingMatches, setIsFetchingMatches] = useState(false);
  const [fetchError, setFetchError] = useState<string | null>(null);

  useEffect(() => {
    if (!open) return;
    
    const fetchMatches = async () => {
      setIsFetchingMatches(true);
      setFetchError(null);
      
      try {
        const { data, error } = await supabase.functions.invoke('pandascore-matches');
        
        if (error) {
          setFetchError('Could not load matches from PandaScore');
          setMatches([]);
          return;
        }
        
        if (data?.error) {
          setFetchError('Could not load matches from PandaScore');
          setMatches([]);
          return;
        }
        
        if (data?.matches && Array.isArray(data.matches)) {
          const parsedMatches = data.matches.map((m: any) => ({
            ...m,
            scheduledAt: m.scheduledAt,
          }));
          setMatches(parsedMatches);
        } else {
          setMatches([]);
        }
      } catch (err) {
        setFetchError('Could not load matches from PandaScore');
        setMatches([]);
      } finally {
        setIsFetchingMatches(false);
      }
    };

    fetchMatches();
  }, [open]);

  const filteredMatches = useMemo(() => {
    if (!searchQuery.trim()) return matches;
    
    const lowerQuery = searchQuery.toLowerCase();
    return matches.filter(match => 
      match.team1.toLowerCase().includes(lowerQuery) ||
      match.team2.toLowerCase().includes(lowerQuery) ||
      match.team1Short.toLowerCase().includes(lowerQuery) ||
      match.team2Short.toLowerCase().includes(lowerQuery) ||
      match.tournament.toLowerCase().includes(lowerQuery)
    );
  }, [searchQuery, matches]);

  useEffect(() => {
    if (selectedMatch) {
      const matchStart = new Date(selectedMatch.scheduledAt);
      const estimatedEnd = addUTCHours(matchStart, 3);
      setResolutionDate(estimatedEnd);
    }
  }, [selectedMatch]);

  const autoGeneratedTitle = useMemo(() => {
    if (!selectedMatch || !resolutionDate) return '';
    const dateLabel = formatUTCDateShort(resolutionDate);
    return `${selectedMatch.team1Short} vs ${selectedMatch.team2Short} Winner [${dateLabel} UTC]`;
  }, [selectedMatch, resolutionDate]);

  const handleCreateClick = () => {
    if (!selectedMatch || !resolutionDate) return;
    setShowConfirmModal(true);
  };

  const handleConfirmedCreate = () => {
    if (!selectedMatch || !resolutionDate) return;
    onCreateFeed({
      game: selectedMatch.game,
      gameName: getGameLabel(selectedMatch.game),
      matchId: selectedMatch.id,
      marketType: 'match_winner',
      resolutionDate,
      title: autoGeneratedTitle,
      // Add team names for better card display
      team1Name: selectedMatch.team1Short,
      team2Name: selectedMatch.team2Short,
    });
    onOpenChange(false);
  };

  const canCreate = connected && selectedMatch && resolutionDate;

  return (
    <Dialog open={open} onOpenChange={onOpenChange}>
      <DialogContent className="sm:max-w-[480px] bg-card border-border">
        <DialogHeader>
          <div className="flex items-center gap-3">
            <img src={esportsIcon} alt="Esports" className="h-10 w-10 object-contain" />
            <div>
              <DialogTitle className="text-xl font-bold">Esports</DialogTitle>
              <p className="text-sm text-muted-foreground">Instant settlement data for global esports tournaments</p>
            </div>
          </div>
        </DialogHeader>
        
        <div className="space-y-4 mt-4">
          {/* Smart Match Finder */}
          <div className="space-y-2">
            <Label className="text-sm font-medium text-foreground">Find Match</Label>
            
            {fetchError && (
              <div className="flex items-center gap-2 p-3 bg-destructive/10 border border-destructive/20 rounded-md">
                <AlertCircle className="h-4 w-4 text-destructive shrink-0" />
                <span className="text-sm text-destructive">{fetchError}</span>
              </div>
            )}
            
            {isFetchingMatches && (
              <div className="flex items-center gap-2 p-3 bg-muted/50 border border-border rounded-md">
                <Loader2 className="h-4 w-4 animate-spin text-muted-foreground" />
                <span className="text-sm text-muted-foreground">Loading matches from PandaScore...</span>
              </div>
            )}
            
            {!isFetchingMatches && !fetchError && (
              <Popover open={matchSearchOpen} onOpenChange={setMatchSearchOpen}>
                <PopoverTrigger asChild>
                  <Button
                    variant="outline"
                    role="combobox"
                    aria-expanded={matchSearchOpen}
                    className="w-full justify-between bg-background border-border h-auto min-h-[42px] py-2"
                  >
                    {selectedMatch ? (
                      <div className="flex items-center gap-2 text-left min-w-0">
                        <span className={cn(
                          "text-[10px] font-bold px-1.5 py-0.5 rounded border shrink-0",
                          getGameBadgeClasses(selectedMatch.game)
                        )}>
                          {getGameLabel(selectedMatch.game)}
                        </span>
                        <div className="flex flex-col items-start min-w-0">
                          <span className="font-medium text-sm truncate max-w-full">
                            {selectedMatch.team1Short} vs {selectedMatch.team2Short}
                          </span>
                          <span className="text-xs text-muted-foreground truncate max-w-full">
                            {format(new Date(selectedMatch.scheduledAt), 'MMM d, HH:mm')} UTC
                          </span>
                        </div>
                      </div>
                    ) : (
                      <span className="text-muted-foreground flex items-center gap-2 min-w-0">
                        <Search className="h-4 w-4 shrink-0" />
                        <span className="truncate">
                          {matches.length > 0 ? "Type team name..." : "No matches"}
                        </span>
                      </span>
                    )}
                    <ChevronsUpDown className="ml-2 h-4 w-4 shrink-0 opacity-50" />
                  </Button>
                </PopoverTrigger>
                <PopoverContent className="w-[350px] p-0 bg-popover border-border z-[60]" align="start">
                  <Command shouldFilter={false}>
                    <CommandInput 
                      placeholder="Search teams..." 
                      value={searchQuery}
                      onValueChange={setSearchQuery}
                    />
                    <CommandList>
                      <CommandEmpty>
                        {matches.length === 0 ? "No upcoming matches found" : "No matches found for your search"}
                      </CommandEmpty>
                      <CommandGroup className="max-h-[280px] overflow-y-auto">
                        {filteredMatches.map((match) => (
                          <CommandItem
                            key={match.id}
                            value={match.id}
                            onSelect={() => {
                              setSelectedMatch(match);
                              setMatchSearchOpen(false);
                              setSearchQuery('');
                            }}
                            className="flex items-start gap-2 cursor-pointer py-3"
                          >
                            <span className={cn(
                              "text-[10px] font-bold px-1.5 py-0.5 rounded border shrink-0 mt-0.5",
                              getGameBadgeClasses(match.game)
                            )}>
                              {getGameLabel(match.game)}
                            </span>
                            <div className="flex-1 min-w-0">
                              <div className="font-medium text-sm">
                                {match.team1} vs {match.team2}
                              </div>
                              <div className="text-xs text-muted-foreground">
                                {match.tournament} • {format(new Date(match.scheduledAt), 'MMM d, HH:mm')} UTC
                              </div>
                            </div>
                            <Check
                              className={cn(
                                "h-4 w-4 shrink-0 mt-0.5",
                                selectedMatch?.id === match.id ? "opacity-100" : "opacity-0"
                              )}
                            />
                          </CommandItem>
                        ))}
                      </CommandGroup>
                    </CommandList>
                  </Command>
                </PopoverContent>
              </Popover>
            )}
            
            {selectedMatch && (
              <p className="text-xs text-muted-foreground">
                Match ID: <span className="font-mono">{selectedMatch.id}</span>
              </p>
            )}
          </div>

          {/* Resolution Date */}
          <div className="space-y-2">
            <Label className="text-sm font-medium text-foreground">
              Resolution Date <span className="text-muted-foreground text-xs">(UTC)</span>
              <span className="text-xs text-muted-foreground ml-2">(auto: match start + 3hrs)</span>
            </Label>
            <div className="flex gap-2">
              <div className={cn(
                "flex-1 flex items-center gap-2 px-3 py-2 rounded-md border bg-muted/50 border-border/50",
                !resolutionDate && "text-muted-foreground"
              )}>
                <CalendarIcon className="h-4 w-4 text-muted-foreground" />
                {resolutionDate ? (
                  <span className="text-sm">{formatUTCDateShort(resolutionDate)}</span>
                ) : (
                  <span className="text-sm text-muted-foreground">Select a match first</span>
                )}
              </div>
              
              <div className="flex items-center gap-1 px-3 py-2 rounded-md border bg-muted/50 border-border/50">
                <Clock className="h-4 w-4 text-muted-foreground" />
                <span className="text-sm font-mono">
                  {resolutionDate ? formatUTCTime(resolutionDate) : '--:--'}
                </span>
                <span className="text-xs text-muted-foreground">UTC</span>
              </div>
            </div>
            
            {selectedMatch && resolutionDate && (
              <p className="text-xs text-muted-foreground">
                Match starts at {formatUTCTime(new Date(selectedMatch.scheduledAt))} UTC, estimated to settle at {formatUTCTime(resolutionDate)} UTC
              </p>
            )}
          </div>

          {/* Auto-generated Feed Title Preview */}
          {autoGeneratedTitle && (
            <div className="space-y-2">
              <Label className="text-sm font-medium text-muted-foreground">Feed Title (auto)</Label>
              <div className="px-3 py-2 bg-muted/50 rounded-md border border-border/50 h-10 flex items-center">
                <span className="text-sm text-foreground truncate">{autoGeneratedTitle}</span>
              </div>
            </div>
          )}

          <div className="flex items-center justify-between pt-4 border-t border-border">
            <FeeDisplay />
            <Button
              variant={deploymentState === 'success' ? 'default' : 'gold'}
              size="sm"
              onClick={handleCreateClick}
              disabled={!canCreate || isLoading}
              className={deploymentState === 'success' ? 'bg-green-500 hover:bg-green-600' : ''}
            >
              {getButtonText ? getButtonText('Create Oracle') : (isLoading ? 'Creating...' : 'Create Oracle')}
            </Button>
          </div>
        </div>
      </DialogContent>

      <TransactionConfirmModal
        open={showConfirmModal}
        onOpenChange={setShowConfirmModal}
        onConfirm={handleConfirmedCreate}
        feedTitle={autoGeneratedTitle}
      />
    </Dialog>
  );
};
